### DISCUSSION TITLE
Custom geometry made up from several geometries.

### DISCUSSION THREAD

**Discussion by stromhaug-tommy** [GeometryGym/GeometryGymIFC#89] [Q&A] [answered]
Hi again and thank you for helping me out previously.

Below if how i tried to solve the case of making one proxy ofbject from several geometry objects. Ocurrecnces are only an instance of an geometry. I Save it as .stl and pare it to build the faceted geometry. It seems like only the first geometry get present in the final result. Cube is misleading, it could be anything.

        public static DatabaseIfc CreateProxyElement(DatabaseIfc db, IfcBuilding building, ComponentOccurrencesEnumerator occurencences, Inventor.Application inventorApp)
        
        {
            rm = new LIST<IfcRepresentationMap>();
            foreach (ComponentOccurrence compOcc in occurencences){

                if (!compOcc.Suppressed)
                {
                    if (!(compOcc.DefinitionDocumentType == DocumentTypeEnum.kPartDocumentObject))
                    {
                        continue;
                    }

                    PartDocument pd = compOcc.Definition.Document as PartDocument;

                    SaveSTL(pd, inventorApp);

                    List<IfcFace> facesShell = GetFaceaForPart(db);

                    IfcClosedShell cs = new IfcClosedShell(facesShell);

                    IfcFacetedBrep k = new IfcFacetedBrep(cs);

                    IfcShapeRepresentation cubeShape = new IfcShapeRepresentation(k);

                    IfcRepresentationMap representationMap = new IfcRepresentationMap(db.Factory.Origin2dPlace, cubeShape);
                    
                    rm.Add(representationMap);
                    
                }
            }

            IfcBuildingElementProxyType cubeProxy = new IfcBuildingElementProxyType(db, "test", IfcBuildingElementProxyTypeEnum.ELEMENT);
            cubeProxy.Name = "Custom Geometry";
            cubeProxy.Description = "";
            cubeProxy.RepresentationMaps = rm;
            
            db.Context.AddDeclared(cubeProxy);
            IfcElement element = cubeProxy.GenerateMappedItemElement(building, new IfcCartesianTransformationOperator3D(db));
            db.WriteFile(@"D:\temptest2.ifc");

**Comment by jmirtsch** [comment, marked as answer]
Rather than create multiple representation maps, create a single representation map with multiple breps in the shape representation if an element geometry comprises multiple shapes.  See if this makes sense.


			List<IfcFacetedBrep> breps = new List<IfcFacetedBrep>();
			foreach (ComponentOccurrence compOcc in occurencences)
			{

				if (!compOcc.Suppressed)
				{
					if (!(compOcc.DefinitionDocumentType == DocumentTypeEnum.kPartDocumentObject))
					{
						continue;
					}

					PartDocument pd = compOcc.Definition.Document as PartDocument;

					SaveSTL(pd, inventorApp);

					List<IfcFace> facesShell = GetFaceaForPart(db);

					IfcClosedShell cs = new IfcClosedShell(facesShell);

					IfcFacetedBrep k = new IfcFacetedBrep(cs);

					breps.Add(k);
					

				}
			}

			var geometricContext = db.Factory.GeometricRepresentationContext(IfcGeometricRepresentationContext.GeometricContextIdentifier.Model);
			IfcShapeRepresentation cubeShape = new IfcShapeRepresentation(geometricContext, breps, ShapeRepresentationType.Brep);

			IfcRepresentationMap representationMap = new IfcRepresentationMap(db.Factory.Origin2dPlace, cubeShape);

			IfcBuildingElementProxyType cubeProxy = new IfcBuildingElementProxyType(db, "test", IfcBuildingElementProxyTypeEnum.ELEMENT);
			cubeProxy.Name = "Custom Geometry";
			cubeProxy.Description = "";
			cubeProxy.RepresentationMaps.Add(representationMap);

**Reply by stromhaug-tommy** [reply]
Perfect!
